pyubx2/ubxreader.py

```diff
@@ -200,6 +205,47 @@ def _parse_nmea(self, hdr: bytes) -> tuple:
             parsed_data = None
         return (raw_data, parsed_data)
 
+    def _parse_rtcm3(self, hdr: bytes) -> tuple:
+        """
+        Handle any RTCM data in the stream.
+
+        NB: pyubx2 does not currently parse RTCM3 data; we
+        simply read to the end of the RTCM3 message and return
+        the raw data bytes in an <RTCM3()> 'wrapper'.
+
+        RTCM3 message format:
+        +---------+--------+--------+----------------+--------+
+        |  0xd3   | 000000 | length |  data message  |  crc   |
+        +---------+--------+--------+----------------+--------+
+        |<-- 8 -->|<- 6 -->|<- 10 ->|<- length x 8 ->|<- 24 ->|
+
+        :param bytes hdr: first 2 bytes of RTCM header
+        :return: tuple of (raw_data as bytes, parsed_stub as string)
+        :rtype: tuple
+        """
+
+        hdr3 = self._read_bytes(1)
+        size = hdr3[0] | (hdr[1] << 8)
+        payload = self._read_bytes(size)
+        crc = self._read_bytes(3)
+        raw_data = hdr + hdr3 + payload + crc
+        parsed_stub = f"<RTCM3(data={payload}, crc={crc})>"

```

Comment:
not sure embedding the payload here is very useful - it is long, and readily accessible from the raw_data variable. What might be useful here is the RTCM message type, extractable as (payload[0] << 4 | payload[1] >> 4)

Reply:
Agreed. I only had to hand an old (2006) copy of the RTCM 10403.1 standard and wasn't sure if the message type was _always_ the first 12 bits of the payload, but assuming it is then fine.

As an aside,  this old copy of the standard also states...

"The Preamble is a fixed 8-bit sequence. The next six bits are reserved in RTCM10403.1 and should be set to zero by the Transport Layer Control for all messages. _The mobile user receiver should ignore these bits and not assume they will always be set to zero_. In future versions these bits may contain the version number of the standard."

... which was partly why I'd avoided attempting to parse RTCM messages in the past, but it works for the time being at least.

I've added a stub `RTCMMessage` class definition as this facilitates the consistent use of message identity and protocol filters, etc.